test-runner/wasi_test_runner/test_executor.py

```diff
@@ -0,0 +1,86 @@
+import glob
+import json
+import os
+import re
+import time
+
+from datetime import datetime
+from typing import List, cast
+
+from .runtime_adapter import RuntimeAdapter
+from .test_case import (
+    TestResult,
+    TestConfig,
+    TestOutput,
+    TestCase,
+)
+from .test_reporters import TestReporter
+from .test_suite import TestSuite
+from .validators import Validator
+
+
+class TestSuiteExecutor:  # pylint: disable=too-few-public-methods
+    def __init__(
+        self, test_suite_path: str, runtime: RuntimeAdapter, validators: List[Validator]
+    ) -> None:
+        self._test_suite_path = test_suite_path
+        self._runtime = runtime
+        self._validators = validators
+        self._test_suite_name = self._read_manifest()
+
+    def run(self, reporter: TestReporter) -> TestSuite:
+        test_cases: List[TestCase] = []
+
+        test_start = datetime.now()
+        for test_path in glob.glob(os.path.join(self._test_suite_path, "*.wasm")):
+            test_case = self._execute_single_test(test_path)
+            test_cases.append(test_case)
+            reporter.report_test(test_case)
+        elapsed = (datetime.now() - test_start).total_seconds()
```

Comment:
The execution time for individual test is ~10ms (for the tests we currently have) - it's more than one can expect mainly because it also includes WASM runtime startup time. My approach is to pass whatever data we have to the reporter, and let it decide what to do with that. I don't have strong opinion whether we should have seconds here or convert it to milliseconds, but I guess it doesn't matter - reporter will do what's right for their purpose.

Reply:
Agreed, what is the precision of this number? I was just wondering whether `total_seconds` would round the 10ms down to 0s, rather than 0.010s. If it's the latter, that's OK and means the reporter can format it how it likes.

Reply:
`total_seconds` has microsecond precision, unless the intervals are significantly large (> 270 years). See [doc](https://docs.python.org/3/library/datetime.html#datetime.timedelta.total_seconds) for details.